src/main/java/org/optaplanner/springboottaskassigning/solver/SolverTask.java

```diff
@@ -16,34 +16,51 @@
 
 package org.optaplanner.springboottaskassigning.solver;
 
+import java.util.function.Consumer;
+
 import org.optaplanner.core.api.score.Score;
 import org.optaplanner.core.api.solver.Solver;
+import org.optaplanner.core.api.solver.event.SolverEventListener;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 public class SolverTask<Solution_> implements Runnable {
 
     private static final Logger logger = LoggerFactory.getLogger(SolverTask.class);
 
-    private final Comparable<?> tenantId;
+    private final Object problemId;
     private Solver<Solution_> solver;
     private Solution_ planningProblem;
+    private Consumer<Solution_> onSolvingEnded;
 
-    public SolverTask(Comparable<?> tenantId, Solver<Solution_> solver, Solution_ planningProblem) {
-        this.tenantId = tenantId;
+    public SolverTask(Object problemId, Solver<Solution_> solver, Solution_ planningProblem,
+                      Consumer<Solution_> onSolvingEnded) {
+        this.problemId = problemId;
         this.solver = solver;
         this.planningProblem = planningProblem;
+        this.onSolvingEnded = onSolvingEnded;
     }
 
     @Override
     public void run() {
-        logger.info("Running solverTask for tenantId ({}).", tenantId);
-        solver.solve(planningProblem);
-        logger.info("Done");
+        try {
+            logger.info("Running solverTask for problemId ({}).", problemId);
+            solver.solve(planningProblem);
+            if (onSolvingEnded != null) {
+                // TODO implement this on a separate thread
+                onSolvingEnded.accept(solver.getBestSolution());
+            }
+            logger.info("Done");
+        } catch (Exception e) {
+            // TODO: better handling of exception (using Callable? / ExtendedExecuter with afterExecute() method)
+            // TODO propagate exception to SolverManager so it restarts solving for this tenant
+            logger.error("Error in SolverTask", e);
+            Thread.currentThread().interrupt();
```

Comment:
Don't interrupt the thread on an exception, only on an interruption, see code in PartitionedSearch etc.